hail/src/main/scala/is/hail/expr/ir/analyses/SemanticHash.scala

```diff
@@ -0,0 +1,409 @@
+package is.hail.expr.ir.analyses
+
+import is.hail.backend.ExecuteContext
+import is.hail.expr.ir.functions.{TableCalculateNewPartitions, WrappedMatrixToValueFunction}
+import is.hail.expr.ir.{MatrixRangeReader, _}
+import is.hail.io.fs.FS
+import is.hail.io.vcf.MatrixVCFReader
+import is.hail.methods._
+import is.hail.types.virtual._
+import is.hail.utils.{Logging, TreeTraversal, toRichBoolean}
+import org.apache.commons.codec.digest.MurmurHash3
+
+import java.nio.ByteBuffer
+import scala.language.implicitConversions
+
+case object SemanticHash extends Logging {
+  type Type = Int
+  type NullableType = Integer
+
+  // Picked from https://softwareengineering.stackexchange.com/a/145633
+  def extend(x: Type, bytes: Array[Byte]): Type =
+    MurmurHash3.hash32x86(bytes, 0, bytes.length, x)
+
+
+  def apply(ctx: ExecuteContext)(root: BaseIR): Option[Type] =
+    ctx.timer.time("SemanticHash") {
+
+      // Running the algorithm on the name-normalised IR
+      // removes sensitivity to compiler-generated names
+      val nameNormalizedIR = ctx.timer.time("NormalizeNames") {
+        new NormalizeNames(_.toString, allowFreeVariables = true)(root)
```

Comment:
Isn't this always run at the top level? What free variables could there be?

Reply:
This is helpful for testing where I'm just using unbound `ref`s

Reply:
I plan to extract this and normalise all names once at the top